t/core/ctx_with_params.t

```diff
@@ -0,0 +1,174 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+use t::APISIX 'no_plan';
+
+repeat_each(1);
+no_long_string();
+no_root_location();
+no_shuffle();
+
+our $yaml_config = <<_EOC_;
+apisix:
+    router:
+        http: 'radixtree_uri_with_parameter'
+_EOC_
+
+add_block_preprocessor(sub {
+    my ($block) = @_;
+
+    if (!$block->yaml_config) {
+        $block->set_value("yaml_config", $yaml_config);
+    }
+});
+
+run_tests;
+
+__DATA__
+
+=== TEST 1: sanity
+--- config
+    location /t {
+        content_by_lua_block {
+            local core = require("apisix.core")
+            local ctx = {}
+            core.ctx.set_vars_meta(ctx)
+
+            ngx.say("remote_addr: ", ctx.var["remote_addr"])
+            ngx.say("server_port: ", ctx.var["server_port"])
+        }
+    }
+--- request
+GET /t
+--- response_body
+remote_addr: 127.0.0.1
+server_port: 1984
+
+
+
+=== TEST 2: add route and get `uri_arg_`
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "methods": ["GET"],
+                    "plugins": {
+                        "serverless-pre-function": {
+                            "phase": "access",
+                            "functions" : ["return function() ngx.log(ngx.INFO, \"uri_arg_path: \", ngx.ctx.api_ctx.var.uri_arg_path) end"]
```

Comment:
I think this test case is not appropriate, only the function `core.utils.resolve_var` can cover your newly added code

Reply:
Not sure I agree with this as for example `cookie_` is tested in [ctx.t](https://github.com/apache/apisix/blob/master/t/core/ctx.t) and not in [utils.t](https://github.com/apache/apisix/blob/master/t/core/utils.t).
As far as I understand the reasoning for this it's because in `ctx*.t` tests are ran regarding the integration of nginx/openresty/third party context vars where as in `utils.t` is for testing the vars specific to apisix implementation.

However, please suggest the tests are you feel are missing and I will add them.

Reply:
I think you are right